src/vvwe.rs

```diff
@@ -107,24 +108,43 @@ impl<A: Actor, T: CausalOp<A>> CausalityBarrier<A, T> {
         match op.happens_after() {
             // Dang! we have a happens after relation!
             Some(dot) => {
-                // Let's buffer this operation then.
+                // Have we already seen the operation we depend on?
                 if !self.saw_site_dot(&dot) {
-                    self.buffer.insert(dot, op);
-                    // and do nothing
-                    None
+                    // No, so buffer this operation until we see it
+                    self.buffer.entry(dot)
+                        .or_default()
+                        .push(op);
+                    vec![]
                 } else {
-                    Some(op)
+                    // Yes! Free this operation and all other operations which depend on this one
+                    self.remove_from_buffer(op)
```

Comment:
we need to make sure to check the buffer for any dependent operations whenever we release an operation.